hw/ip/spi_device/data/spi_device.hjson

```diff
@@ -517,7 +517,8 @@
       fields: [
         { bits: "0"
           name: "busy"
-          desc: "BUSY"
+          desc: '''BUSY signal is cleared when CSb is high. SW should read
+            back the register to confirm the value is cleared.'''
```

Comment:
Did you have a situation in mind that motivates software's need for confirmation?

While the BUSY bit is set in the CSR and/or the BUSY bit in the SPI status register has a clearing event pending, passthrough should be disabled. However, either the host should not use this register's value as an indicator that `spi_host` can be used, or the CSR bit should update in the sys_clk domain immediately, right? Does software care when the clearing event "transfers" to the `sck` domain?

Reply:
The update of the register in SYS_CLK domain is guarded by CSb (2FF synced version of). It updates only when CSb is high (set/clear).

So, in order to support SW to clear at anytime regardless of CSb state, HW should maintain two sets of BUSY. One is for the value to be updated, the other is to the indicator of the current BUSY @ SYS_CLK.

Let's think the case when SW clears the BUSY when SPI is active (CSb = 0). If we remove the limitation of CSb high condition, SW may clear the BUSY signal while SPI transaction is in active. Then, the passthrough will be un-blocked right away.

The BUSY @ SPI_CLK will be updated at the next SPI transaction though, as the design only updates BUSY @ SPI_CLK when CSb is asserted (0 -> 1). So, from the host system point of view, it won't raises the race condition.

To support safer operation, the logic gets complicated. I just wanted simpler logic inside, and read back of BUSY would not harm the SW much in my opinion.

Reply:
Hm... I'm confused. I still don't understand why software cares about when the "clear event" hits the Status Register. The host's spi_host is the only one directly observing the BUSY bit in the Status Register, right? :thinking: 

There are three entities here:
1. Software
   - Software is only interested in issuing a "clear the BUSY bit" instruction to the spi_device, and this can be delayed. 
   - The CSR that represents the BUSY bit's state can update immediately as an option, if you want it to be used to indicate whether spi_host is blocked. However, this probably isn't a good use of that bit, so from software's perspective, this bit might as well be write-only.
   - Software generally should not hold itself up by waiting for the host to send another transaction (OK if that is actually the only job it has). That could come at any time, even hours later.
2. OT spi_host / passthrough
   - The passthrough mux should not change its selection while the spi_device has an active transaction.
   - But OT's spi_host _also_ should not be used while a "clear the BUSY bit" is pending. This intention can be tracked by a driver in RAM; software doesn't need to track the change via the CSR.
3. Host
   - Because we need to make changes to bits outside an active transaction (and we're crossing clock domains), we choose to change the Status Register at CSB posedge and specify a minimum CSB inactive time.

In other words, is there any reason software needs to read the BUSY bit from the CSR? It seems like software has done its job with the write.

Reply:
> Software is only interested in issuing a "clear the BUSY bit" instruction to the spi_device, and this can be delayed.
>
This is not true in the current design. It will be more complicated to support this behavior. And I didn't consider that at the beginning. If I wanted to do that. As I mentioned in the previous comment, the HW should maintain two copies of the BUSY (@ SYS_CLK) one for blocking the passthrough, another for the pending event of clearing the BUSY (@ SPI_CLK).

Reply:
Wait... Are you saying that the SW may wait until SPI transaction is completed? Then, yes SW does not have to read back to check if BUSY is cleared.

But I want to emphasize that the HW does not ensure the SW writes to the BUSY actually written.

Reply:
> But I want to emphasize that the HW does not ensure the SW writes to the BUSY actually written.

That's what is probably where my confusion comes. You might want to word the doc to say that, since it's not clear what software's _must_ do (I was trying to ask "Why does software have to spend time confirming that the change happened?" heh).

Something like... "The clearing of the busy bit may not occur immediately. The completion happens at the end of a SPI transaction, and software may read the bit to confirm when the change occurs."

Then we're all good. :)

Reply:
What was slightly confusing was what this statement meant:
```
SW should read back the register to confirm the value is cleared
```

With that construction, it's not totally clear that there is a conditional. It could be that SW should always read the register and wait for the clear to complete. Or it could be that SW should read the register _if it wants_ to observe the clear to complete--And I can't come up with a good reason why we'd really care to do that. It's fun ambiguity! ;)

Reply:
> Uh-oh... I don't want to say this, but both are wrong. HW ignores any updates to the BUSY signal (from SW) when SPI transaction is active (CSb = 0). So, if SW read-back returns 1, SW shall write again when SPI is in idle.

Oh yeah, that doesn't sound ideal. Wouldn't you agree that software should not need to spin on a SPI transaction, retrying a clear until it's accepted? I'd recommend saving that SW intends to clear, then execute at the first opportunity.